### ISSUE TITLE
Options for Base Modelica output

### ISSUE THREAD

**Issue by casella** [OpenModelica/OpenModelica#12333]
As originally discussed in #12279, depending on the specific application, there could be different requirements for Base Modelica output. At one end of the spectrum, if Base Modelica output is fed into a proper compiler, the flattened model could use all the features of the language, such as arrays, records, parameter-binding equations, slicing of arrays, etc. At the other end of the spectrum, some applications may require to read in more heavily preprocessed Base Modelica code, so that a simple parser is enough to handle the model.

As the requirements are mostly orthogonal to each other, they should probably be activated by flags.

### Basic options when instantiating with `--BaseModelica`

These options can be easily carried out by the NF and should be activated by flags

- [ ] 1. Do not evaluate constant expressions
- [X] 2. Evaluate all parameter expressions: `-d=evaluateAllParameters`
- [ ] 3. Evaluate all parameter expressions except some, e.g. for sensitivity analysis or grey-box parameter identification
- [X] 4. Inline functions with annotation(Inline = true): `--frontendInline`
- [ ] 5. Flatten all arrays and equations to scalars: this should be the default behaviour
- [X] 6. Keep all array declarations and array equations as such: `-d=nonfScalarize,arrayConnect,combineSubscripts,evaluateAllParameters,vectorizeBindings` (could be replaced by single `--arrayPreservingFlattening` flag)
- [ ] 7. Move all binding equations for variables into the equation section (could be activated by `--removeVarBindingEquations`)

Option 1. is currently not available, since the OMC frontend constant-evaluates all constant expressions by default. Although this is probably not such a big deal, Base Modelica allows to define constant explicitly, so for some applications it could be useful if one could disable constant-evaluation by the NF and leave those constants in the flat code.

Option 3. is not yet available, discussed in #4960

Option 5. requires #12332 to be resolved.

Option 7. can be useful for a bare-bones Base Modelica importing tool. Binding equations for variables can be useful in Modelica, because they can be changed upon instantiation by means of modifiers; however, there is no instantiation in Base Modelica models, so moving them to the equation sections has no negative side effects. On the other hand, it avoids the need to process binding equations in addition to regular equations.

### Advanced options

These transformations are currently carried out by the backend, but they could be ported into the frontend to facilitate Base Modelica import by tools without a full-fledged symbolic backend, exploiting the capability of the OMC backend

- [ ] 8. Flatten record declarations and record expressions to their scalar counterparts (e.g. Complex operator record declarations and expressions)
- [ ] 9. Basic alias elimination (b = +/-a)
- [ ] 10. Linear alias elimination (b = m*a + q)
- [ ] 11. Index reduction with fixed state selection

Option 8., 9. and 10. are implemented in the backend and could be ported to the frontend

Option 11. basically requires to de-activate all pre-optimizations, run the Pantelides/ASSC algorithms and generate all the required differentiated equations with the backend. Those could be then copied back to the front-end, alongside with an indications of the state derivatives that became dummy derivatives, which should then be replaced in the flat model (e.g. `der(x)` -> $DER_x) 

### Prioritization

@matteodepascali is currently working on the importing of index-1 flattened DAE models into the `Pyomo-DAE` tool, to compute optimal controls for a large Modelica model of innovative zero-emissions thermal power plants. This is a very interesting application that can open an interesting window of opportunity for OMC in a new area of research.

This is very easily accomplished by parsing Base Modelica output with a simple Python script, as long as options 2., 4., 5., and 7. are available. Options 1. and 4. already are, so we should get him 5. and 7. with very high priority.

Later on, 3., 9., 10., and 11. would also be interesting to him and to other practitioners in that field.

The MARCO compiler already exploits 2. and 6. Being a full-fledged compiler, there is no need to do further pre-processing in OMC for it.

**Comment by perost** [comment]
>     * [ ]  1. Do not evaluate constant expressions

We could skip the phase that goes through the model and evaluates all constants/structural parameters to make sure they're evaluated everywhere and not only in some places, but avoiding all evaluation would be very hard since it's spread out all over the compiler and a lot of things rely on things being evaluated.

>     * [ ]  3. Evaluate all parameter expressions except some, e.g. for sensitivity analysis or grey-box parameter identification

Same deal as above basically, but instead of skipping the evaluation phase completely we'd have a table of parameters to skip. But avoiding evaluation altogether might be even harder than the previous case, and it'll likely have a performance penalty even if you don't use the feature.

>     * [ ]  5. Flatten all arrays and equations to scalars: this should be the default behaviour

In theory this should be fairly simple, but maybe not in practice. I only tried it briefly earlier, and it looked like we got some duplicated variables in the flat model. So there's at least some issues to resolve there.

>     * [ ]  7. Move all binding equations for variables into the equation section (could be activated by `--removeVarBindingEquations`)

This should be easy, we already do this in many cases. But I'm not sure what we want to do with parameters. They'd have to all be changed to be non-fixed, which I think can cause issues, or otherwise we'd have to keep their binding equations which means the tool has to support binding equations anyway. And we can of course not do it for constants if the option to not evaluate constants is used.

**Comment by casella** [comment]
> > * [ ]  1. Do not evaluate constant expressions
> > ```
> 
> We could skip the phase that goes through the model and evaluates all constants/structural parameters to make sure they're evaluated everywhere and not only in some places, but avoiding all evaluation would be very hard since it's spread out all over the compiler and a lot of things rely on things being evaluated.

Sure. First thing coming to my mind, conditionally instantiated components. But I guess 1. is really low priority, unless somebody is asking for it because of some specific reason - at the moment I can't think of any.

> > * [ ]  3. Evaluate all parameter expressions except some, e.g. for sensitivity analysis or grey-box parameter identification
> > ```
> 
> Same deal as above basically, but instead of skipping the evaluation phase completely we'd have a table of parameters to skip. But avoiding evaluation altogether might be even harder than the previous case, and it'll likely have a performance penalty even if you don't use the feature.

For the time being, we can definitely restrict 3. to non-structural parameters, i.e. those design parameters (lenghts, masses, temperatures, pressures) that may be subject to design optimization. Does this help?

> > * [ ]  5. Flatten all arrays and equations to scalars: this should be the default behaviour
> > ```
> 
> In theory this should be fairly simple, but maybe not in practice. I only tried it briefly earlier, and it looked like we got some duplicated variables in the flat model. So there's at least some issues to resolve there.

OK. Contrary to the previous two issues, this has very high priority at the moment.

> > * [ ]  7. Move all binding equations for variables into the equation section (could be activated by `--removeVarBindingEquations`)
> > ```
> 
> This should be easy, we already do this in many cases. But I'm not sure what we want to do with parameters.

Evaluate them to literal constants (option 2.), so that the bindings are trivial to handle; in fact, in that case can actually skip the parameter declarations entirely, since the equation bodies do not contain any parameter, just literal constant expressions and variables.

The idea is to get a flat model whose structure is as simple as possible, so basically you'd always use 7. combined with 2.

> We could just restrict the use of 7. They'd have to all be changed to be non-fixed, which I think can cause issues, or otherwise we'd have to keep their binding equations which means the tool has to support binding equations anyway.

See above, in fact you can just ignore them.

**Comment by casella** [comment]
I'd like to re-open the discussion on points 9. and 10.

For our optimization applications with Modelica2Pyomo, we get the flat model from the Base Modelica output of OMC. It would be nice to have alias elimination applied to it, possibly even linear alias elimination.

I understand this is not trivial, because the current alias elimination algorithms work on the backend data structures, not on the AST from which Base Modelica output is generated. But maybe porting those algorithms is not too complicated. Ideally, I would like to have the equation and initial equation sections to only use the selected representative variable for each alias set, then have a final section in the variable declaration where all the alias variables are defined, with binding equations expressing their relation to the selected representative variable.

@perost, @phannebohm how do you see that?

**Comment by phannebohm** [comment]
Would it be reasonable to try outputting Base Modelica from the backend? It sounds to me that we would need to basically port the whole backend to the frontend, if not now latest when you want index reduction.

**Comment by perost** [comment]
> Would it be reasonable to try outputting Base Modelica from the backend? It sounds to me that we would need to basically port the whole backend to the frontend, if not now latest when you want index reduction.

Maybe it's possible to implement some simplified version of Base Modelica output in the backend that doesn't support all the options that the frontend does, but it's likely a lot of work regardless. And having two separate Base Modelica output implementations is maybe not the best idea.

**Comment by phannebohm** [comment]
@perost I'm looking at `SimCodeMain.translateModel()`. If the backend returns a map from alias crefs to their bindings, could you use that to change `flatModel` and generate `NFFlatString` only after that? Or would it be too much work rewriting `runFrontEndWorkNF`?

**Comment by perost** [comment]
> [@perost](https://github.com/perost) I'm looking at `SimCodeMain.translateModel()`. If the backend returns a map from alias crefs to their bindings, could you use that to change `flatModel` and generate `NFFlatString` only after that? Or would it be too much work rewriting `runFrontEndWorkNF`?

Not sure, we currently dump Base Modelica during the instantiation, not after it. Applying a mapping from SimCode might not be entirely trivial, and dumping the Base Modelica after the frontend will introduce changes compared to what we dump right now.

**Comment by casella** [comment]
In principle, being able to dump BaseModelica from the backend would give us a lot more flexibility, because we could apply any already implemented structural analysis and symbolic manipulation method by just calling it.

In practice, as listed above, there are really only a few backend methods that are really of interest for flat, unsorted models:

1. inline functions (the frontend can do that already)
2. evaluate parameters (the frontend can do that already)
3. flatten records to scalars
4. alias elimination and linear alias elimination
5. index reduction with fixed state selection

One option to get 3. to 5. without porting the algorithms, which is an ugly and dumb work, could be the following:

- flatten the model, with all binding equations on variable declarations moved to the equation section, and generate the corresponding BaseModelica output
- pass the model to the backend
- apply the required backend modules
- back-port the results onto the flat BaseModelica model

For example, the output of the alias elimination module is a list of alias sets, with a representative variable for each one of them and indication of the relations between the representative variable and the aliased ones. I think it would be trivial to parse it and apply it to the flat Base Modelica code. One just needs to 
- replace all occurrences of aliased variable names (which are all quoted, so very easy to spot in the source code) with their representative variable names and the appropriate coefficient, e.g. `'a.b'` ->`(-'a.c')`;
- move the original declarations of aliased variables to the end of the declaration section before the equation section, adding the appropriate binding to the representative variable.

Piece of cake. I would probably be able to program this myself in a couple of days.

Also in the case of index reduction, the outcome of the backend module is a set of dummy derivatives that should replace der() operators and a set of differentiated equations that must be added to the model. Re-inserting them into the flat BaseModelica code shouldn't be too hard.

@perost, @phannebohm, what do you think?
